### DISCUSSION TITLE
ERESOLVE installation problem

### DISCUSSION THREAD

**Discussion by Mario2280** [flyskywhy/react-native-gcanvas#53] [Q&A] [answered]
Hello i try to use react-native-pixi in my simple androind app. I use nodejs 18.12.1
I use next commands:
npx create-expo-app AwesomeProject
cd AwesomeProject
npm install @flyskywhy/react-native-browser-polyfill @flyskywhy/react-native-gcanvas
at this point i get the error:
npm ERR! code ERESOLVE
npm ERR! ERESOLVE unable to resolve dependency tree
npm ERR!
npm ERR! While resolving: awesomeproject@1.0.0
npm ERR! Found: react@18.1.0
npm ERR! node_modules/react
npm ERR!   react@"18.1.0" from the root project
npm ERR!
npm ERR! Could not resolve dependency:
npm ERR! peer react@"^16.8.6" from @flyskywhy/react-native-browser-polyfill@1.0.8
npm ERR! node_modules/@flyskywhy/react-native-browser-polyfill
npm ERR!   @flyskywhy/react-native-browser-polyfill@"*" from the root project
npm ERR!   @flyskywhy/react-native-browser-polyfill@"^1.0.8" from @flyskywhy/react-native-gcanvas@2.3.33
npm ERR!   node_modules/@flyskywhy/react-native-gcanvas
npm ERR!     @flyskywhy/react-native-gcanvas@"*" from the root project
npm ERR!     1 more (@flyskywhy/react-native-browser-polyfill)
npm ERR!
npm ERR! Fix the upstream dependency conflict, or retry
npm ERR! this command with --force, or --legacy-peer-deps
npm ERR! to accept an incorrect (and potentially broken) dependency resolution.
I can't figure out what I'm doing wrong. When i use legasy-peer-deps
I got error
Unable to resolve "abab" from "node_modules\@flyskywhy\react-native-gcanvas\packages\gcanvas\src\context\webgl\RenderingContext.js"gContext.js"

**Comment by flyskywhy** [comment]
Please `npm install abab` first.

Thanks for reminding this 😝 , I'll add "abab" into my `package.json` later.

**Comment by Mario2280** [comment]
@flyskywhy Yes but

nvm use 18.12.1
npx create-expo-app PixiCanvasInReactNative
cd PixiCanvasInReactNative
npm install @flyskywhy/react-native-gcanvas --save
cd .\node_modules\@flyskywhy\react-native-gcanvas\packages\gcanvas\
npm i abab
Then i add your strings in settings.gradle. I found two such files, I don’t know which one I need, so I added them to both(Screen 1)
![image](https://user-images.githubusercontent.com/51860597/204394191-86b4268c-45c4-4169-bae3-541e041d55ea.png)
The same with file react-native.config.js(3 files I put text in each)
![image](https://user-images.githubusercontent.com/51860597/204394659-8c3ec40a-56b7-473a-ab0d-37216870f801.png)
After this I try to run to make sure everything works by using command npm run android
``` js 
import {  View } from 'react-native';
import {GCanvasView} from '@flyskywhy/react-native-gcanvas';

export default function App() {
  return (
    <View>
    <GCanvasView/>
    </View>
  );
}
```
And after building I got this error(and after rebuild too)
![image](https://user-images.githubusercontent.com/51860597/204406594-264e96aa-3f4d-47c7-bb91-20a655adad6a.png)
I hope that you will help me launch this in expo, because many people are now starting development from it

**Comment by flyskywhy** [comment]
Please ref to https://github.com/flyskywhy/snakeRN/commit/33530faf26c8bc3b9d882718619a272e093ec88f

**Reply by Mario2280** [reply]
@flyskywhy sorry, but it doesn't work, I tried to install this repository too, but I ran into a bunch of errors that I tried to fix and ended up stuck. Tried to use 8 and 11 Java. When trying to run npm run android
Task :expo-permissions:compileDebugKotlin FAILED
Task :android:adapters:gcanvas_imageloader_fresco:javaPreCompileDebug FAILED
It seems to me that it will be difficult for people to use the repository you wrote following the installation instructions. Especially if as a result it is impossible to start. Сould you update the installation instructions and indicate which build tools and environment variables you used?

**Reply by flyskywhy** [reply]
v3.1.1 of https://github.com/flyskywhy/snakeRN fix the build error here, you can try v3.1.1

PS: for I also fix a `TypeError: context.save is not a function. (In 'context.save()', 'context.save' is undefined)` in a dependency `"@flyskywhy/react-native-browser-polyfill": "^1.0.4"` of `@flyskywhy/react-native-gcanvas`, so you should remove `node_modules/` and `npm install` all automatically again, or, just `npm install @flyskywhy/react-native-browser-polyfil@1.0.4` or `npm install @flyskywhy/react-native-browser-polyfil@1.0.9` manually.

**Comment by Mario2280** [comment, marked as answer]
@flyskywhy Ok, this time I did everything you said above, also during the build I had to add the local.properties file to avoid the 
```
SDK location not found error. Define location with sdk.dir in the local.properties file or with an ANDROID_HOME environment variable
```
and set environment variable ANDROID_HOME.
Also I used NDK 20.1.5948944 and installed some stuff
![image](https://user-images.githubusercontent.com/51860597/205449147-39883376-e002-4dcc-9d76-fbbae99ac298.png)
Then i got this
![image](https://user-images.githubusercontent.com/51860597/205449277-c740ab0a-c61d-4f25-b921-db21fea03c68.png)
**BUT**
When i opened your app my device show this error

```
error: TypeError: Cannot read properties of undefined (reading 'transformFile')
    at E:\snakeRN\node_modules\metro\src\Bundler.js:91:34
    at Generator.next (<anonymous>)
    at asyncGeneratorStep (E:\snakeRN\node_modules\metro\src\Bundler.js:14:24)
    at _next (E:\snakeRN\node_modules\metro\src\Bundler.js:34:9)
    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
[Sat Dec 03 2022 18:28:29.375]  BUNDLE  ./index.android.js

error: TypeError: Cannot read properties of undefined (reading 'transformFile')
    at E:\snakeRN\node_modules\metro\src\Bundler.js:91:34
    at Generator.next (<anonymous>)
    at asyncGeneratorStep (E:\snakeRN\node_modules\metro\src\Bundler.js:14:24)
    at _next (E:\snakeRN\node_modules\metro\src\Bundler.js:34:9)
    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
[Sat Dec 03 2022 18:29:27.387]  BUNDLE  ./index.android.js

error: TypeError: Cannot read properties of undefined (reading 'transformFile')
    at E:\snakeRN\node_modules\metro\src\Bundler.js:91:34
    at Generator.next (<anonymous>)
    at asyncGeneratorStep (E:\snakeRN\node_modules\metro\src\Bundler.js:14:24)
    at _next (E:\snakeRN\node_modules\metro\src\Bundler.js:34:9)
    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)

```

I log arguments with which an error occurs
```
░░░░░░░░░░░░░░░░ 0.0% (0/1)"E:\\snakeRN\\index.android.js" <-filePath
 {
  "customTransformOptions": {},
  "dev": true,
  "hot": true,
  "inlinePlatform": true,
  "minify": false,
  "platform": "android",
  "experimentalImportSupport": false,
  "unstable_disableES6Transforms": false,
  "nonInlinedRequires": [
    "React",
    "react",
    "react-native"
  ],
  "type": "module",
  "inlineRequires": false
} <- transformOptions

```

Sometimes when I try to rebuild the project, instead of this error, I get a white screen with a green text above "Reloading" or "loading from localhost:8081"

**Reply by flyskywhy** [reply]
Seems a node version (or your mac is M1?) related issue, the solution maybe ref to https://github.com/facebook/react-native/issues/33107#issuecomment-1042612307

**Reply by Mario2280** [reply]
@flyskywhy No I dont using Mac. I try downgrade a node version to 17.3.0 and use --openssl-legacy-provider. It did not help. I got this 
![image](https://user-images.githubusercontent.com/51860597/205467730-920f3ed6-d89f-40e5-970f-530f45fff3ef.png)

**Reply by flyskywhy** [reply]
With this "Unable to load script", I think you miss the one of RN debug steps: `adb reverse tcp:8081 tcp:8081`

**Reply by Mario2280** [reply]
@flyskywhy Crap if i use adb reverse tcp:8081 tcp:8081; node node_modules/react-native/local-cli/cli.js run-android --no-jetifier I got error from previous screen :cry::cry::cry:
![image](https://user-images.githubusercontent.com/51860597/205469213-daec7cb6-cfed-49dc-a070-0286b2b47fdd.png)

**Reply by flyskywhy** [reply]
Please use node v10.22.1 for now which I'm using, I will try node 17.3.0 later to debug some.

**Reply by Mario2280** [reply]
@flyskywhy Ok I overshot a bit and am using 10.24.1. Oh my god, this is finally working. Great!!!!
I also wanted to ask, I'm just a little newbie, the old version of the node does not restrict me from using your library with the latest react-native? Also can i use the newest version of pixi.js and all his features or 4.7 is the ceiling?

**Reply by flyskywhy** [reply]
Software stuffs like node, react-native, pixi or any others are growing their version, no one can tracing all versions persionally, that's why opensource community like github make sense, e.g. with your feed back, I know that latest node can't work with my snakeRN again and I will try debug or desribed it in `README.md`. So you can try newest version of pixi.js and tell me the result 😋

**Reply by Mario2280** [reply]
@flyskywhy Fine! I hope you can offer to update this library for the latest version of react-native and nodejs support

**Reply by flyskywhy** [reply]
> @flyskywhy Ok I overshot a bit and am using 10.24.1. Oh my god, this is finally working. Great!!!! I also wanted to ask, I'm just a little newbie, the old version of the node does not restrict me from using your library with the latest react-native? Also can i use the newest version of pixi.js and all his features or 4.7 is the ceiling?

Breaking: Any APP now can be based on `pixi.js` `@4`, `@5`, `@6` or `@7` directly without `react-native-pixi` as described in [react-native-pixi@3](https://github.com/flyskywhy/react-native-pixi), you can easily follow the comments in example https://github.com/flyskywhy/GCanvasRNExamples/blob/master/app/components/Pixi.js to use any version of pixi.
